core/src/sigverify_stage.rs

```diff
@@ -161,34 +161,27 @@ impl SigVerifyStage {
         Self { thread_hdl }
     }
 
-    pub fn discard_excess_packets(batches: &mut Vec<PacketBatch>, max_packets: usize) {
-        let mut received_ips = HashMap::new();
-        for (batch_index, batch) in batches.iter().enumerate() {
-            for (packet_index, packets) in batch.packets.iter().enumerate() {
-                let e = received_ips
-                    .entry(packets.meta.addr().ip())
-                    .or_insert_with(VecDeque::new);
-                e.push_back((batch_index, packet_index));
-            }
+    pub fn discard_excess_packets(batches: &mut [PacketBatch], mut max_packets: usize) {
+        // Group packets by their incoming IP address.
+        let mut addrs = batches
+            .iter_mut()
+            .rev()
+            .flat_map(|batch| batch.packets.iter_mut().rev())
+            .map(|packet| (packet.meta.addr, packet))
+            .into_group_map();
+        // Allocate max_packets evenly across addresses.
+        while max_packets > 0 && !addrs.is_empty() {
+            let num_addrs = addrs.len();
+            addrs.retain(|_, packets| {
+                let cap = (max_packets + num_addrs - 1) / num_addrs;
+                max_packets -= packets.len().min(cap);
```

Comment:
Couldn't this underflow?

Reply:
I don't think so. Do you have a particular example?